### PR TITLE
Dispatch Index ops to Series

### PR DESCRIPTION
**Pull request by jbrockmendel**
Getting ever closer to a single implementation.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Looks reasonable. What's the ideal final state? For future mico-optimization, I could imagine an implementation that does index-specific things (e.g. deferring for `op(Index, Series)`) and then just `box(op(array1, array2))`. It's a bit unfortunate that we have to pay the Series constructor performance tax on each op. I

**Comment by jbrockmendel** [comment]
> What's the ideal final state?

- Contingent on 2D EA happening, the ops are defined directly on PandasArray and either Index._data is a PandasArray or calls PandasArray(self._data) in the op instead of Series(self)
- Otherwise, it'll take some closure-rearrangement, but I think we can get a version of those ops that operates directly on ndarray.
- Regardless: a single implementation so we get internal consistency for free (not-yet-conformant: RangeIndex, IntegerArray, SparseArray)

File: pandas/core/ops/__init__.py [hunk 1]
```diff
@@ -1702,10 +1703,16 @@ def wrapper(left, right):
             #  does inference in the case where `result` has object-dtype.
             return construct_result(left, result, index=left.index, name=res_name)
 
+        elif isinstance(right, (ABCDatetimeArray, pd.DatetimeIndex)):
+            result = op(left._values, right)
+            return construct_result(left, result, index=left.index, name=res_name)
+
         lvalues = left.values
         rvalues = right
         if isinstance(rvalues, ABCSeries):
```
Comment (jreback) [inline pandas/core/ops/__init__.py]: these cases are the same, yes? 

``isinstance(rvalues, (ABCSeries, ABCIndexClass))``

File: pandas/core/ops/missing.py [hunk 2]
```diff
@@ -148,40 +148,26 @@ def mask_zero_div_zero(x, y, result):
...
+    Call fill_zeros with the appropriate fill value depending on the operation,
+    with special logic for divmod and rdivmod.
 
     Parameters
     ----------
     op : function (operator.add, operator.div, ...)
-    left : object (Index for non-reversed ops)
-    right : object (Index fof reversed ops)
+    left : object (np.ndarray for non-reversed ops)
+    right : object (np.ndarray fof reversed ops)
```
Comment (jreback) [inline pandas/core/ops/missing.py]: fof -> for

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, minor comments ping on green. note still want to move things out of pandas/core/ops/__init__.py to a named module.

**Review by jreback — APPROVED** [review]

File: pandas/core/ops/missing.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/ops/missing.py]: fixed+green

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**